File: src/dev/wirlie/bungeecord/glist/groups/GroupManager.java
package dev.wirlie.bungeecord.glist.groups;

import dev.wirlie.bungeecord.glist.EnhancedBCL;
import net.md_5.bungee.api.connection.ProxiedPlayer;
import net.md_5.bungee.config.Configuration;

import java.io.File;
import java.io.FileInputStream;
import java.io.IOException;
import java.io.InputStreamReader;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.util.*;
import java.util.logging.Level;
import java.util.stream.Collectors;

public class GroupManager {

	private final List<Group> groups = new ArrayList<>();

	private final File groupFile;
	private final EnhancedBCL plugin;

	public GroupManager(EnhancedBCL plugin) {
		this.groupFile = new File(plugin.getDataFolder(), "GroupsPrefix.yml");
		this.plugin = plugin;
	}

	public void loadGroups() {
		if(!groupFile.exists()) {
			plugin.getLogger().info("File not found: GroupsPrefix.yml | Generating file...");
			File parentFolder = groupFile.getParentFile();
			if(!parentFolder.exists()) {
				parentFolder.mkdirs();
			}
			try {
				Files.copy(getClass().getResourceAsStream("/GroupsPrefix.yml"), groupFile.toPath());
			} catch (IOException e) {
				e.printStackTrace();
			}
		}

		if(!groupFile.exists()) {
			plugin.getLogger().warning("File: GroupsPrefix.yml not found, check folder permissions, plugin cannot create default file...");
			return;
		}

		plugin.getLogger().info("GroupsPrefix.yml found, reading configuration ...");

		try {
			groups.clear();

			Configuration config = plugin.getYamlProvider().load(new InputStreamReader(new FileInputStream(groupFile), StandardCharsets.UTF_8));
			String staffGroupPermission = config.getString("staff-group.permission", null);

			if(staffGroupPermission == null) {
				plugin.getLogger().warning("No permission defined for staff members. Using default permission: 'ebl.groups.staff'");
				staffGroupPermission = "ebl.groups.staff";
			}

			if(!config.contains("groups")) {
				plugin.getLogger().warning("No groups defined. Adding default group for users.");
				Group group = new Group("default");
				group.setDefaultGroup(true);
				group.setNameColor("&7");
				groups.add(group);
				return;
			}

			try {
				@SuppressWarnings("unchecked")
				List<Map<?, ?>> allData = (List<Map<?, ?>>) config.get("groups");
				for(Map<?,?> groupData : allData) {
					try {
						String id = null;
						int weight = 0;
						String prefix = null;
						String prefixPermission = null;
						boolean prefixVisibleToUsers = true;
						String nameColor = "";
						boolean defaultGroup = false;

						for (Map.Entry<?, ?> entry : groupData.entrySet()) {
							String key = entry.getKey().toString();
							String value = entry.getValue().toString();

							if (key.equalsIgnoreCase("id")) {
								id = value;
							} else if (key.equalsIgnoreCase("weight")) {
								try {
									weight = Integer.parseInt(value);
								} catch (NumberFormatException e) {
									plugin.getLogger().warning("Unexpected weight value: " + value);
								}
							} else if (key.equalsIgnoreCase("prefix")) {
								prefix = value;
							} else if (key.equalsIgnoreCase("prefix-permission")) {
								prefixPermission = value;
							} else if(key.equalsIgnoreCase("prefix-visible-to-users")) {
								prefixVisibleToUsers = Boolean.parseBoolean(value);
							} else if(key.equalsIgnoreCase("name-color")) {
								nameColor = value;
							} else if(key.equalsIgnoreCase("default-group")) {
								try {
									defaultGroup = Boolean.parseBoolean(value);
								} catch (Exception ex) {
									ex.printStackTrace();
								}
							}
						}

						if(id == null) {
							plugin.getLogger().warning("Group not loaded: id must be defined");
						} else {
							Group group = new Group(id);
							group.setDefaultGroup(defaultGroup);
							group.setWeight(weight);
							group.setPrefix(prefix);
							group.setPrefixPermission(prefixPermission);
							group.setVisibleToUsers(prefixVisibleToUsers);
							group.setNameColor(nameColor);

							if(groups.contains(group)) {
								plugin.getLogger().warning("Group '" + group.getId() + "' already loaded! Fixing id to: '" + group.getId() + "~1'");
								group.setId(id + "~1");
							}

							int fixedId = 1;
							while(groups.contains(group)) {
								fixedId++;
								plugin.getLogger().warning("Group '" + group.getId() + "' already loaded! Fixing id to: '" + id + "~" + fixedId + "'");
								group.setId(id + "~" + fixedId);
							}

							plugin.getLogger().info("Group '" + group.getId() + "' loaded. Weight: " + group.getWeight() + ", permission: " + group.getPrefixPermission());

							groups.add(group);
						}
					} catch (Exception e) {
						e.printStackTrace();
					}
				}
			} catch (ClassCastException e) {
				plugin.getLogger().log(Level.SEVERE, "Cannot read GroupsPrefix.yml (ClassCastException)", e);
			}

			groups.sort(Comparator.naturalOrder());
			plugin.getLogger().info(groups.size() + " groups loaded.");
			plugin.getLogger().info(groups.stream().map(Group::getId).collect(Collectors.joining(", ", "[", "]")));
		} catch (IOException e) {
			plugin.getLogger().log(Level.SEVERE, "Cannot read GroupsPrefix.yml (IOException)", e);
		}
	}

	public Optional<Group> getGroup(ProxiedPlayer player) {
		Group defaultGroup = null;

		for(Group group : groups) {
			if(group.isDefaultGroup()) {
				defaultGroup = group;
				continue;
			}

			if(group.getPrefixPermission() == null || group.getPrefixPermission().isEmpty() || player.hasPermission(group.getPrefixPermission())) {
				return Optional.of(group);
			}
		}

		return Optional.ofNullable(defaultGroup);
	}
}

File: src/main/kotlin/dev/wirlie/bungeecord/glist/groups/GroupManager.kt
package dev.wirlie.bungeecord.glist.groups

import dev.wirlie.bungeecord.glist.EnhancedBCL
import net.md_5.bungee.api.connection.ProxiedPlayer
import java.io.File
import java.io.FileInputStream
import java.io.IOException
import java.io.InputStreamReader
import java.nio.charset.StandardCharsets
import java.nio.file.Files
import java.util.*
import java.util.logging.Level
import java.util.stream.Collectors
import kotlin.ClassCastException
import kotlin.Exception
import kotlin.NumberFormatException
import kotlin.String
import kotlin.toString

class GroupManager(plugin: EnhancedBCL) {
    private val groups: MutableList<Group> = ArrayList()
    private val groupFile: File
    private val plugin: EnhancedBCL

    init {
        groupFile = File(plugin.dataFolder, "GroupsPrefix.yml")
        this.plugin = plugin
    }

    fun loadGroups() {
        if (!groupFile.exists()) {
            plugin.logger.info("File not found: GroupsPrefix.yml | Generating file...")
            val parentFolder = groupFile.parentFile
            if (!parentFolder.exists()) {
                parentFolder.mkdirs()
            }
            try {
                Files.copy(javaClass.getResourceAsStream("/GroupsPrefix.yml")!!, groupFile.toPath())
            } catch (e: IOException) {
                e.printStackTrace()
            }
        }
        if (!groupFile.exists()) {
            plugin.logger.warning("File: GroupsPrefix.yml not found, check folder permissions, plugin cannot create default file...")
            return
        }

        plugin.logger.info("GroupsPrefix.yml found, reading configuration ...")
        try {
            groups.clear()
            val config =
                plugin.yamlProvider!!.load(InputStreamReader(FileInputStream(groupFile), StandardCharsets.UTF_8))

            //TODO: Esta variable no se usa, ¿por que?
            var staffGroupPermission = config.getString("staff-group.permission", null)

            if (staffGroupPermission == null) {
                plugin.logger.warning("No permission defined for staff members. Using default permission: 'ebl.groups.staff'")
                staffGroupPermission = "ebl.groups.staff"
            }

            if (!config.contains("groups")) {
                plugin.logger.warning("No groups defined. Adding default group for users.")
                val group = Group("default")
                group.isDefaultGroup = true
                group.nameColor = "&7"
                groups.add(group)
                return
            }

            try {
                @Suppress("UNCHECKED_CAST")
                val allData = config["groups"] as List<Map<*, *>>

                for (groupData in allData) {
                    try {
                        var id: String? = null
                        var weight = 0
                        var prefix: String? = null
                        var prefixPermission: String? = null
                        var prefixVisibleToUsers = true
                        var nameColor = ""
                        var defaultGroup = false

                        for ((key1, value1) in groupData) {
                            val key = key1.toString()
                            val value = value1.toString()
                            if (key.equals("id", ignoreCase = true)) {
                                id = value
                            } else if (key.equals("weight", ignoreCase = true)) {
                                try {
                                    weight = value.toInt()
                                } catch (e: NumberFormatException) {
                                    plugin.logger.warning("Unexpected weight value: $value")
                                }
                            } else if (key.equals("prefix", ignoreCase = true)) {
                                prefix = value
                            } else if (key.equals("prefix-permission", ignoreCase = true)) {
                                prefixPermission = value
                            } else if (key.equals("prefix-visible-to-users", ignoreCase = true)) {
                                prefixVisibleToUsers = value.toBoolean()
                            } else if (key.equals("name-color", ignoreCase = true)) {
                                nameColor = value
                            } else if (key.equals("default-group", ignoreCase = true)) {
                                try {
                                    defaultGroup = value.toBoolean()
                                } catch (ex: Exception) {
                                    ex.printStackTrace()
                                }
                            }
                        }

                        if (id == null) {
                            plugin.logger.warning("Group not loaded: id must be defined")
                        } else {
                            val group = Group(id)
                            group.isDefaultGroup = defaultGroup
                            group.weight = weight
                            group.prefix = prefix
                            group.prefixPermission = prefixPermission
                            group.isVisibleToUsers = prefixVisibleToUsers
                            group.nameColor = nameColor
                            if (groups.contains(group)) {
                                plugin.logger.warning("Group '" + group.id + "' already loaded! Fixing id to: '" + group.id + "~1'")
                                group.id = "$id~1"
                            }
                            var fixedId = 1
                            while (groups.contains(group)) {
                                fixedId++
                                plugin.logger.warning("Group '" + group.id + "' already loaded! Fixing id to: '" + id + "~" + fixedId + "'")
                                group.id = "$id~$fixedId"
                            }
                            plugin.logger.info("Group '" + group.id + "' loaded. Weight: " + group.weight + ", permission: " + group.prefixPermission)
                            groups.add(group)
                        }
                    } catch (e: Exception) {
                        e.printStackTrace()
                    }
                }
            } catch (e: ClassCastException) {
                plugin.logger.log(Level.SEVERE, "Cannot read GroupsPrefix.yml (ClassCastException)", e)
            }

            groups.sortBy { it }
            plugin.logger.info(groups.size.toString() + " groups loaded.")
            plugin.logger.info(groups.stream().map { obj: Group -> obj.id }
                .collect(Collectors.joining(", ", "[", "]")))
        } catch (e: IOException) {
            plugin.logger.log(Level.SEVERE, "Cannot read GroupsPrefix.yml (IOException)", e)
        }
    }

    fun getGroup(player: ProxiedPlayer): Optional<Group> {
        var defaultGroup: Group? = null
        for (group in groups) {
            if (group.isDefaultGroup) {
                defaultGroup = group
                continue
            }

            if (group.prefixPermission == null || group.prefixPermission!!.isEmpty() || player.hasPermission(group.prefixPermission)) {
                return Optional.of(group)
            }
        }
        return Optional.ofNullable(defaultGroup)
    }
}
